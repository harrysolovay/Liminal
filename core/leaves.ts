import { Ty } from "./Ty.ts"

export type none<P extends string = string> = Ty<null, P>
export const none: Ty<null, never> = Ty((description) => ({
  type: "null",
  description,
}))

export type bool<P extends string = string> = Ty<boolean, P>
export const bool: Ty<boolean, never> = Ty((description) => ({
  type: "boolean",
  description,
}))

export type num<P extends string = string> = Ty<number, P>
export const num: Ty<number, never> = Ty((description) => ({
  type: "number",
  description,
}))

export type str<P extends string = string> = Ty<string, P>
export const str: Ty<string, never> = Ty((description) => ({
  type: "string",
  description,
}))
