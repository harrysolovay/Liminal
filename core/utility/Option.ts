import * as I from "../intrinsics/mod.ts"
import type { Type } from "../Type.ts"

export function Option<V, E>(some: Type<V, E>): Type<V | null, E> {
  return I.union(I.null, some)
}
