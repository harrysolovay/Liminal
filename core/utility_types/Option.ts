import type { Type } from "../Type.ts"
import * as I from "../types/mod.ts"

export function Option<V, E>(some: Type<V, E>): Type<V | null, E> {
  return I.union(I.null, some)
}
