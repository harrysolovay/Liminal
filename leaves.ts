import { Ty } from "./Ty.ts"

export const none = Ty<null>((description) => ({
  type: "null",
  description,
}))

export const boolean = Ty<boolean>((description) => ({
  type: "boolean",
  description,
}))

export const number = Ty<number>((description) => ({
  type: "number",
  description,
}))

export const string = Ty<string>((description) => ({
  type: "string",
  description,
}))
