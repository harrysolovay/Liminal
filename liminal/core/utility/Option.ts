import { null as null_, union } from "../intrinsics/mod.ts"
import type { JSONTypeName } from "../JSONSchema.ts"
import type { Type } from "../Type.ts"

export function Option<V, P extends symbol>(
  some: Type<JSONTypeName, V, P>,
): Type<"union", V | null, P> {
  return union(null_, some)
}
