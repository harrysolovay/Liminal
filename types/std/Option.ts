import { none } from "../primitives.ts"
import { taggedUnion } from "../taggedUnion.ts"
import type { Ty } from "../Ty.ts"

export function option<S extends Ty>(Some: S): ReturnType<
  typeof taggedUnion<{
    Some: S
    None: Ty<null, never>
  }>
> {
  return taggedUnion({
    Some,
    None: none,
  })`An option of specified \`some\` type.`
}
