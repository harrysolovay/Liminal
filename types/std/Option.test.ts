import { option } from "./Option.ts"
import { assertTySnapshot } from "test_util"
import { string } from "../leaves.ts"

Deno.test("Option", async (t) => {
  await assertTySnapshot(t, option(string))
})
