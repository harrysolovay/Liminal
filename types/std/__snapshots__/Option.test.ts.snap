export const snapshot = {};

snapshot[`Option 1`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      anyOf: [
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "Some",
              type: "string",
            },
            value: {
              type: "string",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "None",
              type: "string",
            },
            value: {
              type: "null",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
      ],
      description: "An option of specified \`some\` type.",
      discriminator: "type",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`Option 2`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      anyOf: [
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "Some",
              type: "string",
            },
            value: {
              type: "string",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "None",
              type: "string",
            },
            value: {
              type: "null",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
      ],
      description: "Description one. An option of specified \`some\` type.",
      discriminator: "type",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`Option 3`] = `
{
  additionalProperties: false,
  properties: {
    value: {
      anyOf: [
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "Some",
              type: "string",
            },
            value: {
              type: "string",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "None",
              type: "string",
            },
            value: {
              type: "null",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
      ],
      description: "Description two. Description one. An option of specified \`some\` type.",
      discriminator: "type",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;
